src/Data/Tree/Binary/Zipper/Properties.agda

```diff
@@ -0,0 +1,106 @@
+open import Data.Maybe.Relation.Unary.All using (All; just; nothing)
+open import Relation.Binary.PropositionalEquality
+open ≡-Reasoning
+open import Function
+open import Level using (Level)
+
+private
+  variable
+    a b c d : Level
+    A : Set a
+    B : Set b
+    C : Set c
+    D : Set d
+
+-- Invariant: Zipper represents a given tree
+
+-- Stability under moving
+
+toTree-up-identity : (zp : Zipper A B) → All ((_≡_ on toTree) zp) (up zp)
+toTree-up-identity (mkZipper [] foc) = nothing
+toTree-up-identity (mkZipper (leftBranch m l ∷ ctx) foc) = just refl
+toTree-up-identity (mkZipper (rightBranch m r ∷ ctx) foc) = just refl
+
+toTree-left-identity : (zp : Zipper A B) → All ((_≡_ on toTree) zp) (left zp)
+toTree-left-identity (mkZipper ctx (leaf x)) = nothing
+toTree-left-identity (mkZipper ctx (node l m r)) = just refl
+
+toTree-right-identity : (zp : Zipper A B) → All ((_≡_ on toTree) zp) (right zp)
+toTree-right-identity (mkZipper ctx (leaf x)) = nothing
+toTree-right-identity (mkZipper ctx (node l m r)) = just refl
+
+-- Tree-like operations indeed correspond to their counterparts
+------------------------------------------------------------------------
+
+toTree-size-commute : ∀ (zp : Zipper A B) → size zp ≡ BT.size (toTree zp)
+toTree-size-commute (mkZipper c v) = helper c v
+  where
+    helper : (cs : List (Crumb A B))
+           → (t : Tree A B)
+           → size (mkZipper cs t) ≡ BT.size (toTree (mkZipper cs t))
+    helper [] foc = +-identityʳ (BT.size foc)
+    helper (leftBranch m l ∷ ctx) foc = begin
+      BT.size foc + (1 + (BT.size l + sum (List.map (suc ∘ BT.size ∘ getTree) ctx))) ≡˘⟨ +-assoc (BT.size foc) 1 (BT.size l + sum (List.map (suc ∘ BT.size ∘ getTree) ctx)) ⟩
+      BT.size foc + 1 + (BT.size l + sum (List.map (suc ∘ BT.size ∘ getTree) ctx)) ≡⟨ cong (λ y → y + (BT.size l + sum (List.map (suc ∘ BT.size ∘ getTree) ctx))) (+-comm (BT.size foc) 1) ⟩
+      1 + BT.size foc + (BT.size l + sum (List.map (suc ∘ BT.size ∘ getTree) ctx)) ≡˘⟨ +-assoc (1 + BT.size foc) (BT.size l) (sum (List.map (suc ∘ BT.size ∘ getTree) ctx)) ⟩
+      1 + BT.size foc + BT.size l + sum (List.map (suc ∘ BT.size ∘ getTree) ctx) ≡⟨ cong (λ y → y + sum (List.map (suc ∘ BT.size ∘ getTree) ctx)) (+-comm (1 + BT.size foc) (BT.size l)) ⟩
+      size (mkZipper ctx (node l m foc)) ≡⟨ helper ctx (node l m foc) ⟩
+      BT.size (toTree (mkZipper (leftBranch m l ∷ ctx) foc)) ∎
+    helper (rightBranch m r ∷ ctx) foc = begin
+      BT.size foc + (1 + (BT.size r + sum (List.map (suc ∘ BT.size ∘ getTree) ctx))) ≡˘⟨ cong (λ y → BT.size foc + y) (+-assoc 1 (BT.size r) (sum (List.map (λ x → suc (BT.size (getTree x))) ctx))) ⟩
+      BT.size foc + (1 + BT.size r + sum (List.map (suc ∘ BT.size ∘ getTree) ctx)) ≡˘⟨ +-assoc (BT.size foc) (suc (BT.size r)) (sum (List.map (suc ∘ BT.size ∘ getTree) ctx)) ⟩
+      size (mkZipper ctx (node foc m r)) ≡⟨ helper ctx (node foc m r) ⟩
+      BT.size (toTreeHelper ctx (node foc m r)) ∎
```

Comment:
The subexpressions `sum (List.map (suc ∘ BT.size ∘ getTree) ctx)` appears a lot,
it is worth naming it in a `let`.